pkg/frontend/querymiddleware/shard_active_series.go

```diff
@@ -0,0 +1,277 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querymiddleware
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/go-kit/log"
+	jsoniter "github.com/json-iterator/go"
+	"github.com/opentracing/opentracing-go"
+	otlog "github.com/opentracing/opentracing-go/log"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql/parser"
+	"golang.org/x/sync/errgroup"
+
+	apierror "github.com/grafana/mimir/pkg/api/error"
+	"github.com/grafana/mimir/pkg/distributor"
+	"github.com/grafana/mimir/pkg/querier/stats"
+	"github.com/grafana/mimir/pkg/storage/sharding"
+	"github.com/grafana/mimir/pkg/util"
+	"github.com/grafana/mimir/pkg/util/spanlogger"
+)
+
+type shardActiveSeriesMiddleware struct {
+	upstream http.RoundTripper
+	logger   log.Logger
+}
+
+func newShardActiveSeriesMiddleware(upstream http.RoundTripper, logger log.Logger) http.RoundTripper {
+	return &shardActiveSeriesMiddleware{
+		upstream: upstream,
+		logger:   logger,
+	}
+}
+
+func (s *shardActiveSeriesMiddleware) RoundTrip(r *http.Request) (*http.Response, error) {
+	const defaultNumShards = 1
+
+	ctx := r.Context()
+
+	spanLog, ctx := spanlogger.NewWithLogger(ctx, s.logger, "shardActiveSeries.RoundTrip")
+	defer spanLog.Finish()
+
+	numShards := setShardCountFromHeader(defaultNumShards, r, spanLog)
+
+	if numShards < 2 {
+		spanLog.DebugLog("msg", "query sharding disabled for request")
+		return s.upstream.RoundTrip(r)
+	}
+
+	values, err := util.ParseRequestFormWithoutConsumingBody(r)
+	if err != nil {
+		return nil, apierror.New(apierror.TypeBadData, err.Error())
+	}
+
+	parsed, err := parser.ParseExpr(values.Get("selector"))
+	if err != nil {
+		return nil, apierror.New(apierror.TypeBadData, err.Error())
```

Comment:
We should provide a meaningful response here: which field failed?

Reply:
I added more explicit error messages.